### PR TITLE
feat: add functionality

### PR DESCRIPTION
**Pull request by krlktv**
All tasks are displayed in the corresponding list (ToDo or Completed). 
Ability to create a task. 
Ability to edit the task. 
Ability to delete a task. 
Ability to mark a task as completed (a completed task cannot be edited). 
Sorting by creation date (from new to old and vice versa). 
Assigning a specific color to each task. 
The number of incomplete and completed tasks in parentheses behind the title of the section with tasks. 
Saving tasks (using localStorage).

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+		deleteTask(id);
+		clearForRender();
+		renderTasks();
+	}
+
+	if (e.target.classList.contains('btn-success')) {
+		const id = e.target.closest('.list-group-item').dataset.id;
+		toggleCompleteness(id);
+		clearForRender();
+		renderTasks();
```
Comment (maximsan) [inline app.js:63]: it can be a reusable method

File: app.js [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+	addTaskForm.reset();
+	addTaskForm.querySelector('[data-dismiss="modal"]').click();
+});
+
+tasksContainer.addEventListener('click', e => {
+	if (e.target.classList.contains('btn-danger')) {
+		const id = e.target.closest('.list-group-item').dataset.id;
+		deleteTask(id);
+		clearForRender();
+		renderTasks();
```
Comment (maximsan) [inline app.js:56]: it can be a reusable method

File: app.js [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+		const id = e.target.closest('.list-group-item').dataset.id;
+		toggleCompleteness(id);
+		clearForRender();
+		renderTasks();
+	}
+
+	if (e.target.classList.contains('btn-info')) {
+		const id = e.target.closest('.list-group-item').dataset.id;
+		fillTaskData(id);
+		currentEditId = id;
```
Comment (maximsan) [inline app.js:69]: it can be a reusable method

File: app.js [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+		this.isCompleted = false;
+	}
+}
+
+function getNewTaskData() {
+	const id = localStorage.length;
+	const addTaskTitle = document.querySelector('#inputTitle').value;
+	const addTaskText = document.querySelector('#inputText').value;
+	const addTaskMedium = document.querySelector('#Medium').checked;
+	const addTaskHigh = document.querySelector('#High').checked;
```
Comment (maximsan) [inline app.js:99]: try to use class selectors for such trivial cases

File: app.js [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	let priority = '';
+
+	if (addTaskMedium) {
+		priority = 'Medium';
+	} else if (addTaskHigh) {
+		priority = 'High';
+	} else {
+		priority = 'Low';
```
Comment (maximsan) [inline app.js:107]: create dictionary for priorities

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello. I left some note, let's discuss your implementation later

File: app.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+				color = 'bg-danger';
+			}
+
+			let editBtn = '<button type="button" class="btn btn-info w-100" data-toggle="modal" data-target="#editModal">Edit</button>';
+			if (current.isCompleted) {
+				editBtn = '';
+			}
+
+			const task = document.createElement('li');
+			task.className = `list-group-item d-flex w-100 mb-2 ${color}`;
```
Comment (maximsan) [inline app.js:186]: create method `createTaskClassName`

File: app.js [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+	localStorage.setItem(id, null);
+}
+
+function toggleCompleteness(id) {
+	const task = JSON.parse(localStorage.getItem(id));
+	if (!task.isCompleted) {
+		task.isCompleted = true;
+	} else {
+		task.isCompleted = false;
+	}
```
Comment (maximsan) [inline app.js:236]: use ternary operator

**Review by maximsan — CHANGES_REQUESTED** [review]
I left some more notes. Please review, let me know if you have questions. Thanks!

File: app.js [hunk 8]
```diff
@@ -0,0 +1,268 @@
+'use strict';
+
+const addTaskForm = document.querySelector('#addTaskForm');
+const currentTasks = document.querySelector('#currentTasks');
+const completedTasks = document.querySelector('#completedTasks');
+const incr = document.querySelector('#incr');
```
Comment (Dev1LtJ) [inline app.js:6]: Create semantic names for your variables.

File: app.js [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+const decr = document.querySelector('#decr');
```
Comment (Dev1LtJ) [inline app.js:7]: Create semantic names for your variables.

File: app.js [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+const toDoHeading = document.querySelector('#toDoHeading');
+const completedHeading = document.querySelector('#completedHeading');
+const tasksContainer = document.querySelector('#tasksContainer');
+const clearAll = document.querySelector('#clearAll');
```
Comment (Dev1LtJ) [inline app.js:11]: Create semantic names for your variables.

File: app.js [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+	localStorage.setItem(0, 'incr');
+	clearForRender();
+	renderTasks();
+});
+
+decr.addEventListener('click', () => {
+	localStorage.setItem(0, 'decr');
+	clearForRender();
+	renderTasks();
+});
```
Comment (Dev1LtJ) [inline app.js:32]: Rework it with on method

File: app.js [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+
+clearAll.addEventListener('click', () => {
+	localStorage.clear();
+	localStorage.setItem(0, 'incr');
+	clearForRender();
+	renderTasks();
+});
```
Comment (Dev1LtJ) [inline app.js:39]: What if i want to create new tasks after clearance with reverted sequence?

File: app.js [hunk 13]
```diff
@@ -0,0 +1,268 @@
+'use strict';
```
Comment (Dev1LtJ) [inline app.js:1]: You don't need it, coz you are using class constructor.  It's automatically enables 'use strict'.

File: app.js (hunk 5 above)
Comment (Dev1LtJ) [inline app.js:107]: You can use querySelectorAll and find the checked field.

File: app.js [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+	} else {
+		priority = 'Low';
+	}
+
+	return new Task(id, addTaskTitle, addTaskText, priority);
+}
+
+function addTaskToLocalStorage(task) {
+	localStorage.setItem(task.id, JSON.stringify(task));
+}
```
Comment (Dev1LtJ) [inline app.js:115]: Rework your data flow, as Maxim said.

File: app.js [hunk 15]
```diff
@@ -0,0 +1,268 @@
...
+
+function clearForRender() {
+	while (currentTasks.firstChild) {
+		currentTasks.removeChild(currentTasks.firstChild);
+	}
+
+	while (completedTasks.firstChild) {
+		completedTasks.removeChild(completedTasks.firstChild);
+	}
+}
```
Comment (Dev1LtJ) [inline app.js:125]: I think it's better to rerender only part, that you need.

File: app.js [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+	}
+	return `${hours}:${minutes} ${day}.${month}.${year}`;
+}
+
+function renderTasks() {
+	const tasksArray = [];
+	let toDoCount = 0;
+	let completedCount = 0;
+
+	if (localStorage.getItem(0) === 'incr') {
```
Comment (Dev1LtJ) [inline app.js:158]: Rework your data flow

File: app.js [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+		for (let i = 1; i < localStorage.length; i++) {
+			tasksArray.push(JSON.parse(localStorage.getItem(i)));
```
Comment (Dev1LtJ) [inline app.js:160]: Rework your data flow

File: app.js [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+				<button class="btn btn-secondary h-100" type="button" id="dropdownMenuItem1" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+					<i class="fas fa-ellipsis-v" aria-hidden="true"></i>
+				</button>
+				<div class="dropdown-menu p-2 flex-column" aria-labelledby="dropdownMenuItem1">
+					<button type="button" class="btn btn-success w-100 mb-2">Toggle completeness</button>
+					${editBtn}
+					<button type="button" class="btn btn-danger w-100 mt-2">Delete</button>
+				</div>
+			</div>
+		`;
```
Comment (Dev1LtJ) [inline app.js:210]: I think better way is to create a copy of element, that you have in template and fill it, coz if template will be changed your component wouldn't be match it.

File: app.js [hunk 19]
```diff
@@ -0,0 +1,268 @@
...
+		editModal.querySelector('#editLow').checked = true;
+	} else if (task.priority === 'Medium') {
+		editModal.querySelector('#editMedium').checked = true;
+	} else {
+		editModal.querySelector('#editHigh').checked = true;
+	}
+}
+
+function getEditTaskData() {
+	const task = JSON.parse(localStorage.getItem(currentEditId));
```
Comment (Dev1LtJ) [inline app.js:255]: I think it's bad approach every time to use local storage.

**Review by Dev1LtJ — CHANGES_REQUESTED** [review]
Hello! Excellent work. I left some comments. Try to answer on it all.

File: app.js [hunk 20]
```diff
@@ -0,0 +1,268 @@
...
+const tasksContainer = document.querySelector('#tasksContainer');
+const clearAll = document.querySelector('#clearAll');
+const editTaskForm = document.querySelector('#editTaskForm');
+let currentEditId = '';
+
+clearForRender();
+renderTasks();
+
+if (!localStorage.getItem(0)) {
+	localStorage.setItem(0, 'incr');
```
[the remark this answers is not in the document]
Comment (krlktv) [reply app.js]: Thanks. I fixed it.
Removed the conditional construct.

File: app.js (hunk 1 above)
Comment (krlktv) [reply app.js:61]: Thanks. I refactored it.
```suggestion
		render();
```

File: app.js (hunk 2 above)
Comment (krlktv) [reply app.js:54]: Thanks. I refactored it.
```suggestion
		render();
```

File: app.js (hunk 3 above)
Comment (krlktv) [reply app.js:60]: Thanks. I refactored it.
```suggestion
		const id = getTaskId(e);
		fillTaskData(id);
		currentEditId = id;
```

File: app.js [hunk 21]
```diff
@@ -0,0 +1,268 @@
...
+	if (hours < 10) {
+		hours = '0' + hours;
+	}
+
+	if (minutes < 10) {
+		minutes = '0' + minutes;
+	}
+
+	if (day < 10) {
+		day = '0' + day;
```
[the remark this answers is not in the document]
Comment (krlktv) [reply app.js]: Thanks. I refactored it.
```suggestion
		day = `0${day};
```

File: app.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (krlktv) [reply app.js:148]: Thanks. I refactored it.
```suggestion
	if (localStorage.getItem(0) === SORTING.asc) {
```

File: app.js [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+function renderTasks() {
+	const tasksArray = [];
+	let toDoCount = 0;
+	let completedCount = 0;
+
+	if (localStorage.getItem(0) === 'incr') {
+		for (let i = 1; i < localStorage.length; i++) {
+			tasksArray.push(JSON.parse(localStorage.getItem(i)));
+		}
+	} else if (localStorage.getItem(0) === 'decr') {
```
[the remark this answers is not in the document]
Comment (krlktv) [reply app.js]: Thanks. I refactored it.
```suggestion
	} else if (sortingValue === SORTING.desc) {
```

File: app.js (hunk 5 above)
Comment (krlktv) [reply app.js:101]: Thanks. I fixed it.
```suggestion
		priority = PRIORITY.low;
```

File: app.js (hunk 6 above)
Comment (krlktv) [reply app.js:186]: Thanks. I added it.
```suggestion
			task.className = createTaskClassName(color);
```

File: app.js (hunk 4 above)
Comment (krlktv) [reply app.js:98]: Thank you! 
I will use class selectors for such trivial cases in the future. 
In this case IDs are also needed to bind to the label tag in markup.

File: app.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (krlktv) [reply app.js:210]: Thanks. I fixed it.
```suggestion
		task.innerHTML = createTaskTemplate(current.title, current.priority, current.date, current.text, editBtn);
```

File: app.js [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+			if (current.priority === 'Low') {
+				color = 'bg-light';
+			} else if (current.priority === 'High') {
+				color = 'bg-danger';
+			}
+
+			let editBtn = '<button type="button" class="btn btn-info w-100" data-toggle="modal" data-target="#editModal">Edit</button>';
+			if (current.isCompleted) {
+				editBtn = '';
+			}
```
[the remark this answers is not in the document]
Comment (krlktv) [reply app.js:183]: Thanks. I refactored it.
```suggestion
			const editBtn = !current.isCompleted ? editButton : '';
```

File: app.js (hunk 7 above)
Comment (krlktv) [reply app.js:236]: Thanks. I fixed it.
```suggestion
	task.isCompleted = !task.isCompleted ? true : false;
```

File: app.js (hunk 8 above)
Comment (krlktv) [reply app.js:6]:
```suggestion
const ascBtn = document.querySelector('#incr');
```

File: app.js (hunk 9 above)
Comment (krlktv) [reply app.js:7]: Thanks. I fixed it.
```suggestion
const descBtn = document.querySelector('#decr');
```

File: app.js (hunk 10 above)
Comment (krlktv) [reply app.js:11]: Thanks. I fixed it.
```suggestion
const clearAllBtn = document.querySelector('#clearAll');
```

File: app.js (hunk 12 above)
Comment (krlktv) [reply app.js]: Thanks. I fixed it.
```suggestion
localStorage.setItem(0, sortingValue);
```

File: app.js (hunk 13 above)
Comment (krlktv) [reply app.js:1]: Yes, all code inside the class construct is automatically in strict mode, but not for the whole js file. Thanks!

File: app.js (hunk 15 above)
Comment (krlktv) [reply app.js:125]: Only tasks are rerender here, the rest of the page is not rerender. Thanks!

File: app.js (hunk 18 above)
Comment (krlktv) [reply app.js]: Thanks. I fixed it.
```suggestion
		task.innerHTML = createTaskTemplate(current.title, current.priority, current.date, current.text, editBtn);
```

File: app.js (hunk 11 above)
Comment (krlktv) [reply app.js]: Thanks. I fixed it.
```suggestion
ascBtn.addEventListener('click', () => {
	setSorting(SORTING.asc);
	render();
});

descBtn.addEventListener('click', () => {
	setSorting(SORTING.desc);
	render();
});
```

File: app.js (hunk 16 above)
Comment (krlktv) [reply app.js]: Thanks. I fixed it.
```suggestion
	if (sortingValue === SORTING.asc) {
```

File: app.js (hunk 19 above)
Comment (krlktv) [reply app.js:228]: Thank you!
I will take this fact into account and try to rework the interaction with local storage.

File: app.js (hunk 5 above)
Comment (krlktv) [reply app.js]: Thanks. I'll try to fix it.

File: app.js (hunk 14 above)
Comment (krlktv) [reply app.js:115]: Thanks. I'll try to fix it.

File: index.html [hunk 24]
```diff
@@ -12,6 +12,19 @@
 			min-height: calc(100vh - 15px);
 			margin: 15px auto 0;
 		}
+
+		body.dark {
+			background-color: #434343 !important;
+			color: #fff !important;
```
Comment (maximsan) [inline index.html:18]: don't use `!important`

File: app.js [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+		}
+	} else if (sortingValue === SORTING.desc) {
+		for (let i = localStorage.length - 1; i > 0; i--) {
+			addToTasksArray(tasksArray, i);
+		}
+	}
+
+	for (let i = 0; i < tasksArray.length; i++) {
+		const current = tasksArray[i];
+		if (current === null) {
```
Comment (maximsan) [inline app.js:174]: `if(!current)`

File: app.js [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+
+	if (sortingValue === SORTING.asc) {
+		for (let i = 1; i < localStorage.length; i++) {
+			addToTasksArray(tasksArray, i);
+		}
+	} else if (sortingValue === SORTING.desc) {
+		for (let i = localStorage.length - 1; i > 0; i--) {
+			addToTasksArray(tasksArray, i);
+		}
+	}
```
Comment (maximsan) [inline app.js:170]: create method `sortBy(value)` for this. The method  should be simpler than your implementation

**Review by maximsan — CHANGES_REQUESTED** [review]
It looks better. I left some comments. Let me know if you have questions!

File: app.js (hunk 17 above)
Comment (maximsan) [reply app.js]: @Dev1LtJ please give more details about what you mean. That's not enough to leave a note that it should be reworked. You have to give an insight about what exactly should be changed or propose something or leave a peace of code.

File: app.js (hunk 25 above)
Comment (krlktv) [reply app.js:174]: Thanks. I fixed it.
```suggestion
		if (!current) {
```

File: app.js (hunk 26 above)
Comment (krlktv) [reply app.js:170]: Thanks. I fixed it.
```suggestion
	const tasksArray = sortBy(sortingValue);
```

File: index.html (hunk 24 above)
Comment (krlktv) [reply index.html:18]: Yes, I know this is bad practice. Here I use it to change bootstrap styles.
In the future, I will not use `!important`.

File: app.js [hunk 27]
```diff
@@ -0,0 +1,297 @@
...
+	const addTaskHigh = document.querySelector('#High').checked;
+	let priority = '';
+
+	if (addTaskMedium) {
+		priority = PRIORITY.medium;
+	} else if (addTaskHigh) {
+		priority = PRIORITY.high;
+	} else {
+		priority = PRIORITY.low;
+	}
```
Comment (Dev1LtJ) [inline app.js:111]: rework it with forEach method and take only checked button.

File: app.js (hunk 15 above)
Comment (Dev1LtJ) [reply app.js:128]: You should to rerender only part, that you really need (uncompleted or completed tasks list).

File: app.js [hunk 28]
```diff
@@ -0,0 +1,297 @@
...
+	while (currentTasks.firstChild) {
+		currentTasks.removeChild(currentTasks.firstChild);
+	}
+
+	while (completedTasks.firstChild) {
+		completedTasks.removeChild(completedTasks.firstChild);
+	}
+}
+
+function toTaskDate(taskDate) {
```
Comment (Dev1LtJ) [inline app.js:130]: taskDateToString

File: app.js [hunk 29]
```diff
@@ -0,0 +1,297 @@
...
+
+			if (!current.isCompleted) {
+				currentTasks.appendChild(task);
+				toDoCount++;
+			} else {
+				completedTasks.appendChild(task);
+				completedCount++;
+			}
+		}
+	}
```
Comment (Dev1LtJ) [inline app.js:190]: Rework:
1) Data flow (unnecessary continue);
2) Task priority class. Your priority - is your class.

File: app.js [hunk 30]
```diff
@@ -0,0 +1,297 @@
...
+		} else {
+			let color = 'bg-warning';
+			if (current.priority === PRIORITY.low) {
+				color = 'bg-info';
+			} else if (current.priority === PRIORITY.high) {
+				color = 'bg-danger';
+			}
+
+			const editButton = '<button type="button" class="btn btn-info w-100" data-toggle="modal" data-target="#editModal">Edit</button>';
+			const editBtn = !current.isCompleted ? editButton : '';
```
Comment (Dev1LtJ) [inline app.js:175]: Choice correct name and hide button (do not switch it to empty string).

File: app.js [hunk 31]
```diff
@@ -0,0 +1,297 @@
...
+
+			const task = document.createElement('li');
+			task.className = createTaskClassName(color);
+			task.dataset.id = current.id;
+			task.innerHTML = createTaskTemplate(current.title, current.priority, current.date, current.text, editBtn);
```
Comment (Dev1LtJ) [inline app.js:180]: createTaskHTML

File: app.js [hunk 32]
```diff
@@ -0,0 +1,297 @@
...
+
+function toggleCompleteness(id) {
+	const task = JSON.parse(localStorage.getItem(id));
+	task.isCompleted = !task.isCompleted ? true : false;
+	localStorage.setItem(id, JSON.stringify(task));
+}
+
+function fillTaskData(id) {
+	const task = JSON.parse(localStorage.getItem(id));
+	const editModal = document.querySelector('#editModal');
```
Comment (Dev1LtJ) [inline app.js:208]: Create it with one modal and hide edit or create button

File: app.js [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+	editModal.querySelector('#editInputTitle').value = task.title;
+	editModal.querySelector('#editInputText').value = task.text;
+	if (task.priority === PRIORITY.low) {
+		editModal.querySelector('#editLow').checked = true;
+	} else if (task.priority === PRIORITY.medium) {
+		editModal.querySelector('#editMedium').checked = true;
+	} else {
+		editModal.querySelector('#editHigh').checked = true;
+	}
```
Comment (Dev1LtJ) [inline app.js:217]: Rework it with same name of priority and class (or name or data- or smthg else). Your code will be like: ...querySelector(`${task.priority}`).checked = true;

File: app.js [hunk 34]
```diff
@@ -0,0 +1,297 @@
...
+	task.title = editTaskForm.querySelector('#editInputTitle').value;
+	task.text = editTaskForm.querySelector('#editInputText').value;
+
+	if (editTaskForm.querySelector('#editMedium').checked) {
+		task.priority = PRIORITY.medium;
+	} else if (editTaskForm.querySelector('#editHigh').checked) {
+		task.priority = PRIORITY.high;
+	} else {
+		task.priority = PRIORITY.low;
+	}
```
Comment (Dev1LtJ) [inline app.js:231]: look previous comment.

File: app.js [hunk 35]
```diff
@@ -0,0 +1,297 @@
...
+				<i class="fas fa-ellipsis-v" aria-hidden="true"></i>
+			</button>
+			<div class="dropdown-menu p-2 flex-column" aria-labelledby="dropdownMenuItem1">
+				<button type="button" class="btn btn-success w-100 mb-2">Toggle completeness</button>
+				${editBtn}
+				<button type="button" class="btn btn-danger w-100 mt-2">Delete</button>
+			</div>
+		</div>
+	`;
+}
```
Comment (Dev1LtJ) [inline app.js:273]: As i said, in this case is better to crete copy of an template element, which created for your by author.

File: app.js [hunk 36]
```diff
@@ -0,0 +1,297 @@
...
+
+function setSorting(value) {
```
Comment (Dev1LtJ) [inline app.js:275]: setSortType

**Review by Dev1LtJ — CHANGES_REQUESTED** [review]
Good work! Read my comments and try to fix your data flow.

**Pull request merged**